### PR TITLE
Change Agent.Heuristic to take a float[] argument

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)

Changes the signature of Agent.Heuristic() to take a `float[]` instead of returning one. This eliminate a common source of user error, where the returned array was the wrong size.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

https://jira.unity3d.com/browse/MLA-768
https://github.com/Unity-Technologies/ml-agents/issues/3719

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [x] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [x] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs [hunk 1]
```diff
@@ -31,7 +34,9 @@ public void RequestDecision(AgentInfo info, List<ISensor> sensors)
             StepSensors(sensors);
             if (!info.done)
             {
-                m_LastDecision = m_Heuristic.Invoke();
+                // Reset m_LastDecision each time.
+                 m_LastDecision = new float[m_numActions];
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs:38]: I couldn't decide whether or not to reuse the array here. I figure this approach is safer, and this isn't performance-critical enough to worry about the allocation/GC

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using MLAgents.Sensors;
 using MLAgents.Demonstrations;
 using MLAgents.Policies;
+using UnityEditor;
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Agent.cs:8]: remove

Comment (surfnerd) [reply com.unity.ml-agents/Runtime/Agent.cs:8]:
```suggestion
```

File: com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs (hunk 1 above)
Comment (surfnerd) [reply com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs:38]: I think it depends on the number of agents and the size of the action array.  If there is a large action space and a QA person is trying to make a demo file to use in training but they are hitting GC hitches every few seconds that could make it 1. hard to actually make a good demo. 2. frustrate the hell out of that person.

**Comment by surfnerd** [comment]
overall looks good.  I would rather we re-use the array.  If you feel strongly about no re-using it, let's talk.

File: docs/Migrating.md [hunk 3]
```diff
@@ -15,6 +15,7 @@ The versions can be found in
 * The `play_against_current_self_ratio` self-play trainer hyperparameter has been renamed to `play_against_latest_model_ratio`
 * Removed the multi-agent gym option from the gym wrapper. For multi-agent scenarios, use the [Low Level Python API](Python-API.md).
 * The low level Python API has changed. You can look at the document [Low Level Python API documentation](Python-API.md) for more information. If you use `mlagents-learn` for training, this should be a transparent change.
+* The signature of `Agent.Heuristic()` was changes to take the `float[]` actions as a parameter, instead of returning the array. This was done to prevent a common source of error where users would return arrays of the wrong size.
```
Comment (mmattar) [inline docs/Migrating.md:18]: changes --> changed
the `float[]` --> `float[]` (?)

File: com.unity.ml-agents/Runtime/Agent.cs (hunk 2 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Agent.cs:8]: [shakes fist at Rider]

File: docs/Migrating.md (hunk 3 above)
Comment (chriselion) [reply docs/Migrating.md:18]: Changed to
> The signature of `Agent.Heuristic()` was changed to take a `float[]` as a parameter...
(and in the changelog)

File: com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs (hunk 1 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Policies/HeuristicPolicy.cs:38]: Discussed offline - we need to do a deeper review of array sharing and/or copying around this (and the other policies). Logged as https://jira.unity3d.com/browse/MLA-892 for followup.

**Pull request merged**